balancer-js/src/modules/pools/pool-types/concerns/weighted/exit.concern.ts

```diff
@@ -0,0 +1,309 @@
+   * (ex. [sendTransaction](https://web3js.readthedocs.io/en/v1.2.11/web3-eth.html#sendtransaction)).
+   *
+   * @param {ExitPool}          exitPool - exit pool information to be encoded
+   * @param {string}            exitPool.poolId - id of pool being exited
+   * @param {string}            exitPool.sender - account address sending BPT to exit pool
+   * @param {string}            exitPool.recipient - account address receiving tokens after exiting pool
+   * @param {ExitPoolRequest}   exitPool.exitPoolRequest - object containing exit pool request information
+   * @returns {string}          encodedExitPoolData - Returns an ABI byte string containing the data of the function call on a contract
+   */
+  static encodeExitPool({
+    poolId,
+    sender,
+    recipient,
+    exitPoolRequest,
+  }: ExitPool): string {
+    const vaultInterface = Vault__factory.createInterface();
+
+    return vaultInterface.encodeFunctionData('exitPool', [
+      poolId,
+      sender,
+      recipient,
+      exitPoolRequest,
+    ]);
+  }
+
+  // Exit Concern
+
+  /**
+   * Build exit pool transaction parameters with exact BPT in and minimum token amounts out based on slippage tolerance
+   * @param {string}  exiter - Account address exiting pool
+   * @param {Pool}    pool - Subgraph pool object of pool being exited
+   * @param {string}  bptIn - BPT provided for exiting pool
+   * @param {string}  slippage - Maximum slippage tolerance in percentage. i.e. 0.05 = 5%
+   * @param {string}  singleTokenMaxOut - Optional: token address that if provided will exit to given token
+   * @returns         transaction request ready to send with signer.sendTransaction
+   */
+  buildExitExactBPTIn = ({
+    exiter,
+    pool,
+    bptIn,
+    slippage,
+    singleTokenMaxOut,
+  }: ExitExactBPTInParameters): ExitPoolAttributes => {
+    if (!bptIn.length || parseFixed(bptIn, 18).isNegative()) {
+      throw new BalancerError(BalancerErrorCode.INPUT_OUT_OF_BOUNDS);
+    }
+    if (
+      singleTokenMaxOut &&
+      !pool.tokens.map((t) => t.address).some((a) => a === singleTokenMaxOut)
+    ) {
+      throw new BalancerError(BalancerErrorCode.TOKEN_MISMATCH);
+    }
+
+    const parsedPoolInfo = this.parsePoolInfo(pool); // Parse pool info into EVM amounts in order to match amountsIn scalling
+    const sortedPoolInfo = this.sortPoolInfo(
+      parsedPoolInfo.tokens,
+      parsedPoolInfo.balances,
+      parsedPoolInfo.weights,
+      parsedPoolInfo.decimals
+    );
+
+    let minAmountsOut = Array(parsedPoolInfo.tokens.length).fill('0');
+    let userData: string;
+
+    if (singleTokenMaxOut) {
+      // Exit pool with single token using exact bptIn
+
+      const singleTokenMaxOutIndex =
+        parsedPoolInfo.tokens.indexOf(singleTokenMaxOut);
+
+      const amountOut = SDK.WeightedMath._calcTokenOutGivenExactBptIn(
+        new OldBigNumber(parsedPoolInfo.balances[singleTokenMaxOutIndex]),
```

Comment:
I thought we were moving bignumber.js out of the SDK? I'm ok with keeping it in, I'd probably refactor a lot of my liquidity calculations using it if we are.

Reply:
Yeap, we're moving bignumber.js out of the SDK. We're temporarily relying on georgeroman's lib until John finishes implementing maths on the SOR to replace this.